tests/components/discovergy/test_coordinator.py

```diff
@@ -0,0 +1,67 @@
+"""Test cases for the Discovery update coordinator."""
+
+from unittest.mock import Mock, patch
+
+from pydiscovergy.error import DiscovergyClientError, HTTPError, InvalidLogin
+import pytest
+
+from homeassistant.components.discovergy import (
+    DiscovergyData,
+    DiscovergyUpdateCoordinator,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers.update_coordinator import UpdateFailed
+
+from tests.common import MockConfigEntry
+from tests.components.discovergy.const import LAST_READING
+
+
+@pytest.mark.parametrize(
+    ("sideeffect", "c_error"),
+    [
+        (
+            HTTPError,
+            UpdateFailed,
+        ),
+        (
+            DiscovergyClientError,
+            UpdateFailed,
+        ),
+        (
+            InvalidLogin,
+            ConfigEntryAuthFailed,
+        ),
+    ],
+)
+async def test_coordinator_failed_update(
+    hass: HomeAssistant,
+    mock_config_entry: MockConfigEntry,
+    mock_meters: Mock,
+    sideeffect: str,
+    c_error: str,
+) -> None:
+    """Test the Discovergy coordinator."""
+
+    with patch(
+        "pydiscovergy.Discovergy.meter_last_reading",
+        return_value=LAST_READING,
+    ) as mock_data:
+        await hass.config_entries.async_setup(mock_config_entry.entry_id)
+        await hass.async_block_till_done()
+
+        data: DiscovergyData = hass.data[mock_config_entry.domain][
+            mock_config_entry.entry_id
+        ]
+        coordinator: DiscovergyUpdateCoordinator = data.coordinators[
+            "f8d610b7a8cc4e73939fa33b990ded54"
+        ]
+
+        mock_data.side_effect = sideeffect
+        coordinator.config_entry = None  # config_entry needs to be reset otherwise config flow will be stuck in reauth
+
+        await coordinator.async_refresh()
```

Comment:
Euh, I've seen comments before where was advised not to interact with the coordinator or other internal objects.

Can we maybe test what the effect will be on the sensors? For example, if ConfigEntryAuthFailed is raised, we know that a reauth flow is created so we can test for that. For the update failed we know the sensors will probably be unavailable.

Reply:
Right. Accessing `hass.data` in new tests is not allowed, and should be removed when existing tests are refactored

https://developers.home-assistant.io/docs/development_testing?_highlight=wri#writing-tests-for-integrations
> Make sure to not interact with any integration details in tests of integrations. Following this pattern will make the tests more robust for changes in the integration.

Reply:
Instead of adding tests for the coordinator module, you could consider adding a tests for the `config_flow` (a variant on `test_form_cannot_connect`). And open a new PR for tests on the coordinator.

Reply:
The tests for the coordinator module have been removed, I will add them in another PR.